File: src/core/schema_tests/python/test_hand.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for HandPoseT and related types in isaacteleop.schema.

HandPoseT is a FlatBuffers struct (read-only from Python) that represents hand pose data:
- joints: Direct access to 26 HandJointPose entries via .joint(i) and .num_joints (XR_HAND_JOINT_COUNT_EXT)
- is_active: Whether the hand pose data is active
- timestamp: Timestamp struct with device and common time

HandJointPose is a struct containing:
- pose: The Pose (position and orientation)
- is_valid: Whether this joint data is valid
- radius: The radius of the joint (from OpenXR)

Note: Python code should only READ this data (created by C++ trackers), not modify it.
"""

import pytest

from isaacteleop.schema import (
    HandPoseT,
    HandJointPose,
    Pose,
    Point,
    Quaternion,
)


class TestHandJointPoseConstruction:
    """Tests for HandJointPose construction."""

    def test_default_construction(self):
        """Test default construction creates HandJointPose with default values."""
        joint_pose = HandJointPose()

        assert joint_pose is not None
        # Default pose values should be zero.
        assert joint_pose.pose.position.x == 0.0
        assert joint_pose.pose.position.y == 0.0
        assert joint_pose.pose.position.z == 0.0
        assert joint_pose.is_valid is False
        assert joint_pose.radius == 0.0

    def test_construction_with_values(self):
        """Test construction with position, orientation, is_valid, and radius."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        joint_pose = HandJointPose(pose, True, 0.01)

        assert joint_pose.pose.position.x == pytest.approx(1.0)
        assert joint_pose.pose.position.y == pytest.approx(2.0)
        assert joint_pose.pose.position.z == pytest.approx(3.0)
        assert joint_pose.is_valid is True
        assert joint_pose.radius == pytest.approx(0.01)


class TestHandJointPoseAccess:
    """Tests for HandJointPose property access."""

    def test_pose_access(self):
        """Test accessing pose property."""
        position = Point(1.5, 2.5, 3.5)
        orientation = Quaternion(0.1, 0.2, 0.3, 0.9)
        pose = Pose(position, orientation)
        joint_pose = HandJointPose(pose, True, 0.015)

        assert joint_pose.pose.position.x == pytest.approx(1.5)
        assert joint_pose.pose.orientation.w == pytest.approx(0.9)

    def test_is_valid_access(self):
        """Test accessing is_valid property."""
        pose = Pose(Point(), Quaternion())
        joint_pose = HandJointPose(pose, True, 0.0)

        assert joint_pose.is_valid is True

    def test_radius_access(self):
        """Test accessing radius property."""
        pose = Pose(Point(), Quaternion())
        joint_pose = HandJointPose(pose, False, 0.025)

        assert joint_pose.radius == pytest.approx(0.025)


class TestHandJointPoseRepr:
    """Tests for HandJointPose __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        joint_pose = HandJointPose(pose, True, 0.01)

        repr_str = repr(joint_pose)

        assert "HandJointPose" in repr_str
        assert "Pose" in repr_str


class TestHandPoseTConstruction:
    """Tests for HandPoseT construction and basic properties."""

    def test_default_construction(self):
        """Test default construction creates HandPoseT with zero-initialized fields."""
        hand_pose = HandPoseT()

        assert hand_pose is not None
        assert hand_pose.num_joints == 26
        assert hand_pose.is_active is False
        assert hand_pose.timestamp.device_time == 0


class TestHandPoseTRepr:
    """Tests for HandPoseT __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        hand_pose = HandPoseT()

        repr_str = repr(hand_pose)
        assert "HandPoseT" in repr_str

File: src/core/schema_tests/cpp/test_hand.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated HandPose FlatBuffer struct.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/hand_generated.h>

#include <type_traits>

#define VT(field) (field + 2) * 2
// Only keep Record VT asserts
static_assert(core::HandPoseRecord::VT_DATA == VT(0));
static_assert(std::is_trivially_copyable_v<core::HandPose>);

// =============================================================================
// Compile-time verification of HandJointPose struct.
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::HandJointPose>, "HandJointPose should be a trivially copyable struct");

// =============================================================================
// HandJointPose Tests
// =============================================================================
TEST_CASE("HandJointPose struct can be created and accessed", "[hand][struct]")
{
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.01f);

    SECTION("Pose values are accessible")
    {
        CHECK(joint_pose.pose().position().x() == 1.0f);
        CHECK(joint_pose.pose().position().y() == 2.0f);
        CHECK(joint_pose.pose().position().z() == 3.0f);
    }

    SECTION("is_valid is accessible")
    {
        CHECK(joint_pose.is_valid() == true);
    }

    SECTION("radius is accessible")
    {
        CHECK(joint_pose.radius() == Catch::Approx(0.01f));
    }
}

TEST_CASE("HandJointPose default construction", "[hand][struct]")
{
    core::HandJointPose joint_pose;

    // Default values should be zero/false.
    CHECK(joint_pose.pose().position().x() == 0.0f);
    CHECK(joint_pose.pose().position().y() == 0.0f);
    CHECK(joint_pose.pose().position().z() == 0.0f);
    CHECK(joint_pose.is_valid() == false);
    CHECK(joint_pose.radius() == 0.0f);
}

// =============================================================================
// HandPose Tests
// =============================================================================
TEST_CASE("HandPose can handle is_active value properly", "[hand][native]")
{
    // Create HandPose with default values, is_active should be false.
    core::HandPose hand_pose;
    CHECK(hand_pose.is_active() == false);

    // Set is_active to true.
    hand_pose.mutate_is_active(true);
    CHECK(hand_pose.is_active() == true);
}

TEST_CASE("HandPose default construction", "[hand][native]")
{
    core::HandPose hand_pose;

    // Default values.
    CHECK(hand_pose.is_active() == false);
    // Struct fields are always present (no null checks needed)
}

TEST_CASE("HandPose joints array has correct size", "[hand][native]")
{
    core::HandPose hand_pose;

    // HandPose should have exactly 26 joints (XR_HAND_JOINT_COUNT_EXT).
    CHECK(hand_pose.joints()->size() == 26);
}

TEST_CASE("HandPose can store timestamp", "[hand][native]")
{
    core::HandPose hand_pose;

    // Set timestamp (XrTime is int64_t).
    int64_t test_device_time = 1234567890123456789LL;
    int64_t test_common_time = 9876543210LL;
    core::Timestamp timestamp(test_device_time, test_common_time);
    hand_pose.mutable_timestamp() = timestamp;

    CHECK(hand_pose.timestamp().device_time() == test_device_time);
    CHECK(hand_pose.timestamp().common_time() == test_common_time);
}

TEST_CASE("HandPose joints can be mutated via flatbuffers Array", "[hand][native]")
{
    core::HandPose hand_pose;

    // Create a joint pose.
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.015f);

    // Mutate first joint
    hand_pose.mutable_joints()->Mutate(0, joint_pose);

    // Verify.
    const auto* first_joint = (*hand_pose.joints())[0];
    CHECK(first_joint->pose().position().x() == 1.0f);
    CHECK(first_joint->pose().position().y() == 2.0f);
    CHECK(first_joint->pose().position().z() == 3.0f);
    CHECK(first_joint->is_valid() == true);
    CHECK(first_joint->radius() == Catch::Approx(0.015f));
}

TEST_CASE("HandPose all 26 joints can be set and verified", "[hand][native]")
{
    core::HandPose hand_pose;

    // Set all 26 joints with unique positions.
    for (size_t i = 0; i < 26; ++i)
    {
        core::Point position(static_cast<float>(i), 0.0f, 0.0f);
        core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
        core::Pose pose(position, orientation);
        core::HandJointPose joint_pose(pose, true, 0.01f);
        hand_pose.mutable_joints()->Mutate(i, joint_pose);
    }

    // Verify all joints.
    for (size_t i = 0; i < 26; ++i)
    {
        const auto* joint = (*hand_pose.joints())[i];
        CHECK(joint->pose().position().x() == Catch::Approx(static_cast<float>(i)));
        CHECK(joint->is_valid() == true);
    }
}

TEST_CASE("HandPose can store all fields", "[hand][native]")
{
    core::HandPose hand_pose;

    // Set a few joint poses
    core::Point position(1.5f, 2.5f, 3.5f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.02f);

    hand_pose.mutable_joints()->Mutate(0, joint_pose);
    hand_pose.mutate_is_active(true);
    core::Timestamp timestamp(9876543210LL, 1234567890LL);
    hand_pose.mutable_timestamp() = timestamp;

    // Verify.
    CHECK(hand_pose.joints()->size() == 26);

    const auto* first_joint = (*hand_pose.joints())[0];
    CHECK(first_joint->pose().position().x() == Catch::Approx(1.5f));
    CHECK(first_joint->pose().position().y() == Catch::Approx(2.5f));
    CHECK(first_joint->pose().position().z() == Catch::Approx(3.5f));
    CHECK(first_joint->is_valid() == true);
    CHECK(first_joint->radius() == Catch::Approx(0.02f));

    CHECK(hand_pose.is_active() == true);
    CHECK(hand_pose.timestamp().device_time() == 9876543210LL);
    CHECK(hand_pose.timestamp().common_time() == 1234567890LL);
}

TEST_CASE("HandPose joints can be accessed by index", "[hand][native]")
{
    core::HandPose hand_pose;

    // Access first and last entries (returns pointers).
    const auto* first = (*hand_pose.joints())[0];
    const auto* last = (*hand_pose.joints())[25];

    // Default values should be zero.
    CHECK(first->pose().position().x() == 0.0f);
    CHECK(last->pose().position().x() == 0.0f);
}
